### DISCUSSION TITLE
Support `Unpack` aka PEP 692 annotations

### DISCUSSION THREAD

**Discussion by multimeric** [fastapi/typer#665] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Unpack, TypedDict
import typer

class Args(TypedDict):
    a: int
    b: str

def main(**kwargs: Unpack[Args]):
    pass

typer.run(main)

```
```
Traceback (most recent call last):

  File "/private/var/folders/q5/fvpyppm924525yl9m3xpgjm400031g/T/tmp.dvMKlgyv/test.py", line 11, in <module>
    typer.run(main)

RuntimeError: Type not yet supported: *<class '__main__.Args'>
```

### Description

Above is a fairly simple example of how `Unpack` might be used with typer. There are many reasons why a user might want to do this, [as outlined in the PEP](https://peps.python.org/pep-0692/). My immediate use case for this is to re-use an argument definition I use elsewhere, so I don't have to duplicate the function's type signature and risk the duplicated copies getting out of sync.

Typer would be able to support this by introspecting the `TypedDict` using [`__optional_keys__`](https://docs.python.org/3/library/typing.html#typing.TypedDict.__optional_keys__) etc.

### Operating System

Linux, macOS

### Operating System Details

Monterey (12.6)

### Typer Version

0.9.0

### Python Version

3.11.5

**Comment by elupus** [comment]
This would solve a lot of argument re-use requirements in https://github.com/tiangolo/typer/issues/153

**Comment by nronnei** [comment]
Also running into this, and would love to have it. Would be really nice to be able to define a base set of cli options to include on all methods. There's a _lot_ going on in #153 that would be greatly simplified with PEP 692 support.

**Comment by aantn** [comment]
Strong yes from us! We're using typer for github.com/robusta-dev/holmesgpt/ and github.com/robusta-dev/krr/ and this would solve some annoying problems.

**Comment by aantn** [comment]
Giving this more thought, I believe that TypedDict doesn't allow something like this:

```
class Args(TypedDict):
    a: int = typer.Option(...)
```

However Annotated should work fine,  which is an acceptable compromise I think

**Reply by multimeric** [reply]
You're right that `Annotated` will be necessary, but that's [already the current recommendation for typer]( https://typer.tiangolo.com/release-notes/#features_5:~:text=It%20is%20now%20recommended%20to%20use%20Annotated%20for%20typer.Option()%20and%20typer.Argument()).

**Reply by aantn** [reply]
Yep. We don't use it in our projects today despite the official recommendation because it's lesser known syntax and we don't want to intimidate first time contributors who aren't python experts in favor of something that has no real benefits for us. But being able to reuse arguments between commands would be reason enough to switch. So still +1 from me

**Reply by nronnei** [reply]
I started an attempt at adding support for `TypedDict` & `Unpack` and got a fair bit of the way along, but I'm not good enough in Python to feel confident about my choices. Either of you open to looking at it?

**Reply by multimeric** [reply]
Share the PR and I might be able to have a look.

**Reply by aantn** [reply]
@nronnei any updates? Very interested in trying it

**Reply by nronnei** [reply]
Sorry for the very long delay, @aantn. Here's my very WIP attempt: https://github.com/tiangolo/typer/compare/master...nronnei:typer:nronnei/add-kwargs-typeddict-support

I haven't been able to get back to this since I'd started.

**Comment by mayrholu** [comment]
any updates on this? I would still very much appreciate this feature!

**Comment by LANCER1254** [comment]
I took a look at the current approach for supporting `Unpack[TypedDict]` in kwargs.

From what I can see, the main issue is that `Unpack[...]` is passed all the way down to `get_click_type()`, where it fails with:

    RuntimeError: Type not yet supported: Unpack[...]

I experimented locally and confirmed that Typer currently doesn't expand `**kwargs: Unpack[TypedDict]` before the Click conversion stage.

It seems that the only viable approach is to expand the TypedDict fields into individual keyword-only parameters before ParamMeta generation (e.g. inside `get_params_from_function()`).

I also checked a WIP branch that tries this, and the general direction looks correct, but a few points might need refinement:

- Debug print statements still present
- `Unpack` not explicitly imported
- All fields default to `None`, so `Required` / `NotRequired` are not handled
- `default_factory` handling might override user intent
- Reconstructing `Annotated` may lose typing semantics

Conceptually, I think the key part is mapping:

- Required keys → required CLI options
- NotRequired keys → optional CLI options
- defaults → Click defaults

Would it make sense to explicitly handle `Required` / `NotRequired` here?

Happy to help test or iterate on this further.
